### ISSUE TITLE
SignatureHelp documentation string not escaped in markdown conversion

### ISSUE THREAD

**Issue by hassec** [neovim/neovim#16563]
### Neovim version (nvim -v)

NVIM v0.6.0

### Language server name/version

clangd 13

### Operating system/version

Arch Linux

### Steps to reproduce using "nvim -u minimal_init.lua"

Only changed 2 lines in `minimal_init.lua`: 
```lua
  local name = 'clangd'
  local cmd = { 'clangd'} -- needed for elixirls, omnisharp, sumneko_lua
```
then open a file with 
```cpp
#include<functional>
int main(){
  int a{1};
  int b{2};

  std::swap(a, b);
//             |---- put cursor here and <c-k> for signatureHelp
}
```

**clangd reply**
```
{
  activeParameter = 1,
  activeSignature = 0,
  signatures = { {
      documentation = "\n  @brief Swaps two values.\n  @param  __a  A thing of arbitrary type.\n  @param  __b  Another thing of arbitrary type.\n  @return   Nothing.\n  ",
      label = "swap(_Tp &__a, _Tp &__b) -> typename enable_if<__and_<__not_<__is_tuple_like<_Tp>>, is_move_constructible<_Tp>, is_move_assignable<_Tp>>::value>::type",
      parameters = { {
          label = { 5, 13 }
        }, {
          label = { 15, 23 }
        } }
    } }
}
```

### Expected behavior

I think the above-mentioned `clangd` reply should be handled correctly. 

From looking at the documentation of [`SignatureInformation`](https://microsoft.github.io/language-server-protocol/specifications/specification-3-17/#signatureInformation) the `documentation`  field is allowed to either be `string` or `MarkupContent`

I've locally played around a bit, and the conversion happens here:
https://github.com/neovim/neovim/blob/730a9514dcf540c4921bbe45c1c8849f773bae39/runtime/lua/vim/lsp/util.lua#L818-L820

So one could fix it with the following change:      
```lua
  if signature.documentation then
    if type(signature.documentation) == 'string' then
      signature.documentation = {kind = 'plaintext', value=signature.documentation}
    end
    vim.lsp.util.convert_input_to_markdown_lines(signature.documentation, contents)
  end
```

Which seems to work for my specific case.

![image](https://user-images.githubusercontent.com/11351671/145109061-04413fb3-e035-4eb2-a916-2e4839e69d93.png)

I'm happy to turn this into a PR but wasn't sure if this is maybe a bit too hacky, or if I'm missing something in respect to the LSP spec or it's implementation in neovim, thus figured I'd start with an issue :blush: 

### Actual behavior

The documentation is interpreted as Markdown, thus the double underscores mess with formatting.

![image](https://user-images.githubusercontent.com/11351671/145107987-4bae51cc-051e-4e20-8fb1-2b23f40f6ba1.png)

**Comment by mjlbach** [comment]
I can't reproduce your example, this is what I get when I request signature help at the same position.

```
[DEBUG][2022-01-04 08:16:31] .../vim/lsp/rpc.lua:454	"rpc.receive"	{
  id = 2,
  jsonrpc = "2.0",
  result = {
    activeParameter = 1,
    activeSignature = 0,
    signatures = { {
        documentation = "@endcond",
        label = "swap(_Tp &__a, _Tp &__b) -> typename enable_if<__and_<__not_<__is_tuple_like<_Tp>>, is_move_constructible<_Tp>, is_move_assignable<_Tp>>::value>::type",
        parameters = { {
            label = { 5, 13 }
          }, {
            label = { 15, 23 }
          } }
      } }
  }
}
```

**Comment by hassec** [comment]
Mhh weird..., what version of `clangd` are you using @mjlbach?
I'm on `13.0.0` and just double-checked again, but my log still shows:
```
[DEBUG][2022-01-04 17:27:30] .../vim/lsp/rpc.lua:454	"rpc.receive"	{
  id = 2,
  jsonrpc = "2.0",
  result = {
    activeParameter = 1,
    activeSignature = 0,
    signatures = { {
        documentation = "\n  @brief Swaps two values.\n  @param  __a  A thing of arbitrary type.\n  @param  __b  Another thing of arbitrary type.\n  @return   Nothing.\n  ",
        label = "swap(_Tp &__a, _Tp &__b) -> typename enable_if<__and_<__not_<__is_tuple_like<_Tp>>, is_move_constructible<_Tp>, is_move_assignable<_Tp>>::value>::type",
        parameters = { {
            label = { 5, 13 }
          }, {
            label = { 15, 23 }
          } }
      } }
  }
}
```

**Comment by mjlbach** [comment]
13.0.0
```
❯ clangd --version
clangd version 13.0.0 (Fedora 13.0.0-3.fc35)
Features: linux
Platform: x86_64-unknown-linux-gnu; target=x86_64-redhat-linux-gnu
```

main.cpp

```
#include<functional>

int main(){
  int a{1};
  int b{2};

  std::swap(a, b);
//             |---- put cursor here and <c-k> for signatureHelp
}
```

compile_commands.json
```
[
{
  "directory": "/home/michael/Repositories/test_c_project/build",
  "command": "/usr/bin/c++    -o CMakeFiles/main.dir/main.cpp.o -c /home/michael/Repositories/test_c_project/main.cpp",
  "file": "/home/michael/Repositories/test_c_project/main.cpp"
}
]

```

**Comment by hassec** [comment]
I was simply opening the file and letting `clangd` run in single file mode, but I've just tried the `compile_commands.json` way and that doesn't change anything for me either...  

Any other ideas what could be different?
```
❯ clangd --version
clangd version 13.0.0
Features: linux
Platform: x86_64-pc-linux-gnu
```

**Comment by mjlbach** [comment]
Not really. Can you try in a container and send me the dockerfile? I want to test an alternative to your fix.

**Comment by hassec** [comment]
I can reproduce it with the following docker file:
```
FROM archlinux:latest
Run pacman -Sy --noconfirm neovim git clang
WORKDIR /neovim_test
COPY minimal_init.lua minimal_init.lua
COPY main.cpp main.cpp
```

Assuming that the Dockerfile is in a directory that has the `minimal_init.lua` and `main.cpp` 
I then ran `docker build . -t neovim_tmp` 
and then `docker run -it neovim_tmp`   

If I then open neovim with `nvim -u minimal_init.lua main.cpp` it installs the plugins etc as expected but somehow doesn't correctly launch the LSP, but if I close neovim and then reopen, `clangd` launches, `LspInfo` shows it's running in single file mode, and a `C-k` triggers the same signature help as I've previously observed.  

Sorry for the delay, the docker part needed a bit of googling as it's not something I use daily :sweat_smile:

**Comment by hassec** [comment]
@mjlbach did you have a chance to test the Dockerfile?

**Comment by hassec** [comment]
Gentle reminder @mjlbach. I've just checked, and I can still reproduce the same incorrect behavior on the current master of neovim.  

Fix in #16702 still works, and I've just rebased that PR
